### PR TITLE
chore: support cmd click on all clickable items

### PR DESCRIPTION
**Pull request by YounixM**
## Pull Request

---

Users expect Cmd+Click (Mac) / Ctrl+Click (Windows/Linux) on clickable items to open the target in a new browser tab - a universal web convention. Currently, SPA navigation via `safeNavigate` always replaces the current page, ignoring modifier keys. Several components (e.g. `DashboardsList`, `ListAlert`) had ad-hoc `event.metaKey || event.ctrlKey` checks with `window.open`, leading to duplicated logic and inconsistent coverage.

This PR centralizes modifier-key detection in `useSafeNavigate` by accepting an optional `event` parameter. When a click event is provided and Cmd/Ctrl is held, the hook automatically opens the target in a new tab instead of performing SPA navigation. A shared `utils/navigation.ts` module provides the low-level helpers (`isModifierKeyPressed`, `openInNewTab`, `navigateToPage`) for cases outside the hook.

50 files are updated to thread the mouse event from click handlers into `safeNavigate`, covering: homepage service tables, infra monitoring K8s lists (Pods, Nodes, Clusters, Deployments, etc.), hosts list, alert rules, dashboards list, metrics explorer, side nav, onboarding flows, messaging queues, breadcrumbs, and more. Two files (`DashboardsList`, `ListAlert`) that had duplicated manual Cmd/Ctrl detection were refactored to use the centralized approach.

#### Screenshots / Screen Recordings (if applicable)
N/A - behavioral change: hold Cmd/Ctrl while clicking any navigable row, link, or button to open in a new tab.

#### Issues closed by this PR

---

### Change Type
_Select all that apply_

- [x] Feature
- [ ] Bug fix
- [x] Refactor
- [ ] Infra / Tooling
- [x] Test-only

---

### Bug Context
N/A - this is a new feature / UX enhancement, not a bug fix.

---

- Tests added/updated:
  - `src/utils/__tests__/navigation.test.ts` - 15 tests covering `isModifierKeyPressed`, `openInNewTab`, and `navigateToPage` with various modifier key combinations, edge cases (both metaKey and ctrlKey), and fallback to SPA navigation.
  - `src/hooks/__tests__/useSafeNavigate.test.ts` - 9 tests verifying the mock contract accepts `event` and `newTab` options, and that the `shouldOpenNewTab` decision logic correctly combines explicit `newTab` with modifier-key detection.
- Manual verification: Cmd+Click on service rows, dashboard list items, alert rule links, K8s resource tables, side nav items, and breadcrumbs all open in a new tab. Normal clicks continue to use SPA navigation.
- Edge cases covered: Both `metaKey` (Mac) and `ctrlKey` (Windows/Linux), `newTab: true` without event, event without modifier keys, both `newTab` and modifier keys simultaneously.

---

- Blast radius: Frontend-only. Affects click behavior across ~50 components. No backend changes.
- Potential regressions:
  - A click handler that previously ignored the event parameter now passes it through - if any component has an `onClick` that provides a non-standard event shape, `isModifierKeyPressed` will safely return `false` (checking `metaKey`/`ctrlKey` on undefined returns falsy).
  - Factory functions in `MetricsApplication/Tabs/util.ts` changed return type from `VoidFunction` to `(e?: React.MouseEvent) => void` - callers passing these as `onClick` handlers are compatible since React's `onClick` provides the event as the first argument.
- Rollback plan: Revert the single commit. No data migration or API changes involved.

---

### Changelog

| Field | Value |
|------|-------|
| Deployment Type | Cloud / OSS / Enterprise |
| Change Type | Feature |
| Description | Cmd/Ctrl+Click on navigable items (table rows, links, buttons) now opens the target in a new browser tab instead of replacing the current page. |

---

### Checklist
- [x] Tests added or explicitly not required
- [x] Manually tested
- [x] Breaking changes documented
- [x] Backward compatibility considered

---

## Notes for Reviewers

**Core changes (read first):**
- `src/utils/navigation.ts` - new shared helpers for modifier-key detection
- `src/hooks/useSafeNavigate.ts` - accepts `event` option, auto-detects Cmd/Ctrl
- `__mocks__/useSafeNavigate.ts` - mock updated to match new interface

**Pattern across all 50 files:**
1. Add `React` import (for `React.MouseEvent` type)
2. Thread mouse event from `onClick` handler into `safeNavigate(..., { event })`
3. For table `onRow` handlers: change callback signature to accept `(record, event)` and pass `event` through

**Refactored duplicated logic:**
- `DashboardsList.tsx` - removed manual `event.metaKey || event.ctrlKey` + `window.open` (6 lines to 1 line)
- `ListAlert.tsx` - replaced `openInNewTab: boolean` parameter with `{ event?, newTab? }` options object, removed manual `window.open` branch

**Intentionally not changed** - `safeNavigate` calls inside `useEffect`, `onSuccess` mutation callbacks, `setTimeout`, drag handlers, and pagination `onChange` handlers where no `MouseEvent` is available. These are programmatic navigations where Cmd+Click does not apply.

### CODE REVIEW

File: frontend/src/components/LogDetail/index.tsx [hunk 1]
```diff
@@ -227,7 +227,9 @@ function LogDetailInner({
 				),
 			),
 		};
-		safeNavigate(`${ROUTES.LOGS_EXPLORER}?${createQueryParams(queryParams)}`);
+		safeNavigate(`${ROUTES.LOGS_EXPLORER}?${createQueryParams(queryParams)}`, {
+			event: e,
+		});
```
Comment (manika-signoz) [inline frontend/src/components/LogDetail/index.tsx:240]: safeNavigate has a newTab parameter in options, passing event might not be needed

Comment (YounixM) [reply frontend/src/components/LogDetail/index.tsx:240]: This helps to evaluate the event at one single place rather than validating at every consumption

File: frontend/src/utils/navigation.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
+import React from 'react';
+
+/**
+ * Returns true if the user is holding Cmd (Mac) or Ctrl (Windows/Linux)
+ * during a click event — the universal "open in new tab" modifier.
+ */
+export const isModifierKeyPressed = (
+	event: MouseEvent | React.MouseEvent,
+): boolean => event.metaKey || event.ctrlKey;
```
Comment (H4ad) [inline frontend/src/utils/navigation.ts]: Should add support for middle click with e.button === 1

File: frontend/src/container/InfraMonitoringK8s/Clusters/K8sClustersList.tsx [hunk 3]
```diff
@@ -450,7 +453,19 @@ function K8sClustersList({
...
+		record: K8sClustersRowData,
+		event: React.MouseEvent,
+	): void => {
+		if (event && isModifierKeyPressed(event)) {
+			const newParams = new URLSearchParams(searchParams);
+			newParams.set(
+				INFRA_MONITORING_K8S_PARAMS_KEYS.CLUSTER_NAME,
+				record.clusterUID,
+			);
+			openInNewTab(`${window.location.pathname}?${newParams.toString()}`);
```
Comment (H4ad) [inline frontend/src/container/InfraMonitoringK8s/Clusters/K8sClustersList.tsx]: With a couple repetitions, maybe introducing another helper like `openInNewTabWith(window.location.pathname, searchParams, { INFRA...: record.clusterUId })`

Not blocking

File: frontend/src/hooks/useSafeNavigate.ts [hunk 4]
```diff
@@ -1,11 +1,13 @@
...
+import React, { useCallback } from 'react';
 import { useLocation, useNavigate } from 'react-router-dom-v5-compat';
 import { cloneDeep, isEqual } from 'lodash-es';
+import { isModifierKeyPressed } from 'utils/navigation';
 
 interface NavigateOptions {
 	replace?: boolean;
 	state?: any;
 	newTab?: boolean;
+	event?: MouseEvent | React.MouseEvent;
```
Comment (AshwinBhatkal) [inline frontend/src/hooks/useSafeNavigate.ts]: I don't think we should be adding this event just for this. We already have an option for newTab. So we should populate that option with isModifierKeyPressed function when passing the newTab option.

Comment (YounixM) [reply frontend/src/hooks/useSafeNavigate.ts]: This helps to evaluate the event at one single place rather than validating at every consumption

Comment (AshwinBhatkal) [reply frontend/src/hooks/useSafeNavigate.ts]: The navigation function's job is to do navigation only. It does not need to care about which event got it here. This violates the Single Responsibility Priniciple. 

When navigating, that option should be passed to open in new tab for which the function already has a param. It is okay if we validate at every consumption because we already have a helper for it.

Ideal scenario would be to have a Link component (with some styles based on usage) with path filled in. That should solve it. Do we plan to get there at some point?

Regarding single place validation, if the goal was to ensure that people don't miss it, then it does not solve that as well as the developer still needs to pass in the event object manually. If they miss it, we go back to the previous state where command click does not work. If we did want to make something required, it would be better to make `openInNewTab` required. This makes it explicit and we avoid this issue overall in the future. It is better than passing in the entire event object which is not relevant to the navigation function.

Comment (YounixM) [reply frontend/src/hooks/useSafeNavigate.ts]: Agree @AshwinBhatkal .

**Review by H4ad — APPROVED** [review]

File: frontend/src/container/InfraMonitoringK8s/Nodes/K8sNodesList.tsx [hunk 5]
```diff
@@ -437,7 +438,20 @@ function K8sNodesList({
 		return nodesData.find((node) => node.nodeUID === selectedNodeUID) || null;
 	}, [selectedNodeUID, groupBy.length, nodesData, nestedNodesData]);
 
-	const handleRowClick = (record: K8sNodesRowData): void => {
+	const openNodeInNewTab = (record: K8sNodesRowData): void => {
+		const newParams = new URLSearchParams(searchParams);
+		newParams.set(INFRA_MONITORING_K8S_PARAMS_KEYS.NODE_UID, record.nodeUID);
+		openInNewTab(`${window.location.pathname}?${newParams.toString()}`);
```
Comment (AshwinBhatkal) [inline frontend/src/container/InfraMonitoringK8s/Nodes/K8sNodesList.tsx]: There is a function for this - buildAbsolutePath. All instances same case

File: frontend/src/container/MetricsApplication/Tabs/util.ts [hunk 6]
```diff
@@ -93,8 +93,8 @@ export function onViewTracePopupClick({
 	isViewLogsClicked,
 	stepInterval,
 	safeNavigate,
-}: OnViewTracePopupClickProps): VoidFunction {
-	return (): void => {
+}: OnViewTracePopupClickProps): (e?: React.MouseEvent) => void {
+	return (e?: React.MouseEvent): void => {
```
Comment (AshwinBhatkal) [inline frontend/src/container/MetricsApplication/Tabs/util.ts:97]: can this event actually be undefined?

File: frontend/src/hooks/useSafeNavigate.ts [hunk 7]
```diff
@@ -122,8 +123,9 @@ export const useSafeNavigate = (
 				);
 			}
 
-			// If newTab is true, open in new tab and return early
-			if (options?.newTab) {
+			const shouldOpenNewTab = options?.newTab;
```
Comment (AshwinBhatkal) [inline frontend/src/hooks/useSafeNavigate.ts]: nit: shouldOpenInNewTab

File: frontend/src/utils/navigation.ts [hunk 8]
```diff
@@ -0,0 +1,17 @@
+import React from 'react';
+
+/**
+ * Returns true if the user is holding Cmd (Mac) or Ctrl (Windows/Linux)
+ * during a click event, or if the middle mouse button is used —
+ * the universal "open in new tab" modifiers.
+ */
+export const isModifierKeyPressed = (
```
Comment (AshwinBhatkal) [inline frontend/src/utils/navigation.ts]: Is this file the right place for this?

File: frontend/src/components/LogDetail/index.tsx [hunk 9]
```diff
@@ -1,6 +1,7 @@
-import { useCallback, useEffect, useMemo, useState } from 'react';
+/* eslint-disable sonarjs/cognitive-complexity */
```
Comment (AshwinBhatkal) [inline frontend/src/components/LogDetail/index.tsx]: can we move this comment to the function level?

Comment (YounixM) [reply frontend/src/components/LogDetail/index.tsx]: done

File: frontend/src/utils/navigation.ts (hunk 8 above)
Comment (YounixM) [reply frontend/src/utils/navigation.ts]: moved to app.ts util

File: frontend/src/hooks/useSafeNavigate.ts (hunk 7 above)
Comment (YounixM) [reply frontend/src/hooks/useSafeNavigate.ts]: done

File: frontend/src/container/InfraMonitoringK8s/Nodes/K8sNodesList.tsx (hunk 5 above)
Comment (YounixM) [reply frontend/src/container/InfraMonitoringK8s/Nodes/K8sNodesList.tsx]: done
